sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InsertMapSortInGroupingExpressions.scala

```diff
@@ -34,12 +35,46 @@ object InsertMapSortInGroupingExpressions extends Rule[LogicalPlan] {
     _.containsPattern(AGGREGATE), ruleId) {
     case a @ Aggregate(groupingExpr, _, _) =>
       val newGrouping = groupingExpr.map { expr =>
-        if (!expr.isInstanceOf[MapSort] && expr.dataType.isInstanceOf[MapType]) {
-          MapSort(expr)
+        if (!expr.exists(_.isInstanceOf[MapSort])
+          && expr.dataType.existsRecursively(_.isInstanceOf[MapType])) {
```

Comment:
instead of doing two recursive checks ahead, we can simply skip adding `MapSort` in `insertMapSortRecursively` if the expression is already `MapSort`.